### PR TITLE
feat: add CMake features to build all libraries

### LINKED ISSUE #9333: Create a simple way to compile *all* the libraries
Currently one needs to provide the full list of libraries in `-DGOOGLE_CLOUD_CPP_ENABLE=...` to compile all the libraries. It would be nice to have a special value to compile them all, maybe `*` or simply `all`.

### PR DESCRIPTION
**Pull request by coryan**
Fixes #9333

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10623)

### CODE REVIEW

File: ci/cloudbuild/builds/cmake-install.sh [hunk 1]
```diff
@@ -37,7 +36,7 @@ cmake "${cmake_args[@]}" \
   -DCMAKE_INSTALL_MESSAGE=NEVER \
   -DBUILD_TESTING=OFF \
   -DGOOGLE_CLOUD_CPP_ENABLE_EXAMPLES=OFF \
-  -DGOOGLE_CLOUD_CPP_ENABLE="${ENABLED_FEATURES}"
+  -DGOOGLE_CLOUD_CPP_ENABLE="__ga_libraries__,__experimental_libraries__,experimental-storage-grpc"
```
Comment (dbolduc) [inline ci/cloudbuild/builds/cmake-install.sh:39]: can you add `experimental-opentelemetry` back in?

File: cmake/GoogleCloudCppFeatures.cmake [hunk 2]
```diff
@@ -139,6 +139,17 @@ export_list_to_bazel(
 # Handle the dependencies between features. That is, if feature "X" is enabled
 # also enable feature "Y" because "X" depends on "Y".
 function (google_cloud_cpp_enable_deps)
+    if (__ga_libraries__ IN_LIST GOOGLE_CLOUD_CPP_ENABLE)
+        list(REMOVE_ITEM GOOGLE_CLOUD_CPP_ENABLE __ga_libraries__)
+        list(APPEND GOOGLE_CLOUD_CPP_ENABLE ${GOOGLE_CLOUD_CPP_GA_LIBRARIES})
+        list(APPEND GOOGLE_CLOUD_CPP_ENABLE
+             ${GOOGLE_CLOUD_CPP_TRANSITION_LIBRARIES})
+    endif ()
+    if (__experimental_libraries__ IN_LIST GOOGLE_CLOUD_CPP_ENABLE)
```
Comment (dbolduc) [inline cmake/GoogleCloudCppFeatures.cmake:148]: Are `__ga_libraries__` and `__experimental_libraries__` public? (I am assuming yes, and that the documentation is coming later).

I claim that no one would ever build `GOOGLE_CLOUD_CPP_ENABLE=__experimental_libraries__`.

So I would prefer offering `GOOGLE_CLOUD_CPP_ENABLE=__experimental_and_ga_libraries__` as a way to build all libraries. (I wish we could say `__all_libraries__`, but we must force the customers to type the dreaded letters: "experimental")

File: ci/cloudbuild/builds/shared.sh [hunk 3]
```diff
@@ -24,15 +24,14 @@ source module ci/cloudbuild/builds/lib/features.sh
...
-read -r ENABLED_FEATURES < <(features::list_full_cmake)
 mapfile -t cmake_args < <(cmake::common_args)
 
 INSTALL_PREFIX="/var/tmp/google-cloud-cpp"
 cmake "${cmake_args[@]}" \
   -DCMAKE_INSTALL_PREFIX="${INSTALL_PREFIX}" \
   -DCMAKE_INSTALL_MESSAGE=NEVER \
   -DBUILD_SHARED_LIBS=ON \
-  -DGOOGLE_CLOUD_CPP_ENABLE="${ENABLED_FEATURES}"
+  -DGOOGLE_CLOUD_CPP_ENABLE="__ga_libraries__,__experimental_libraries__"
```
Comment (dbolduc) [inline ci/cloudbuild/builds/shared.sh:34]: are you sure you want to drop: `experimental-opentelemetry` and `experimental-storage-grpc` ?

File: ci/cloudbuild/builds/publish-docs.sh [hunk 4]
```diff
@@ -21,14 +21,13 @@ source module ci/cloudbuild/builds/lib/cmake.sh
...
 mapfile -t FEATURE_LIST < <(features::list_full)
-read -r ENABLED_FEATURES < <(features::list_full_cmake)
 
 version=""
 doc_args=(
   "-DCMAKE_BUILD_TYPE=Debug"
   "-DGOOGLE_CLOUD_CPP_GENERATE_DOXYGEN=ON"
   "-DGOOGLE_CLOUD_CPP_GEN_DOCS_FOR_GOOGLEAPIS_DEV=ON"
-  "-DGOOGLE_CLOUD_CPP_ENABLE=${ENABLED_FEATURES}"
+  "-DGOOGLE_CLOUD_CPP_ENABLE=__ga_libraries__,__experimental_libraries__"
```
Comment (dbolduc) [inline ci/cloudbuild/builds/publish-docs.sh:30]: Can you add: `experimental-opentelemetry` and `experimental-storage-grpc` back in?

I think you will lose documentation such as: https://googleapis.dev/cpp/google-cloud-storage/latest/namespacegoogle_1_1cloud_1_1storage__experimental.html

File: ci/cloudbuild/builds/cmake-single-feature.sh [hunk 5]
```diff
@@ -52,6 +52,11 @@ check_pkgconfig_absolute() {
   fi
 }
 
+for feature in __ga_libraries__ __experimental_libraries__; do
```
Comment (dbolduc) [inline ci/cloudbuild/builds/cmake-single-feature.sh:55]: What does this do that the loop below does not?

**Review by dbolduc — CHANGES_REQUESTED** [review]

File: cmake/GoogleCloudCppFeatures.cmake (hunk 2 above)
Comment (coryan) [reply cmake/GoogleCloudCppFeatures.cmake:148]: Yes, they are public.  I am thinking about where the documentation will go.

File: ci/cloudbuild/builds/cmake-install.sh (hunk 1 above)
Comment (coryan) [reply ci/cloudbuild/builds/cmake-install.sh]: Done

File: cmake/GoogleCloudCppFeatures.cmake (hunk 2 above)
Comment (coryan) [reply cmake/GoogleCloudCppFeatures.cmake:148]: Fixed the documentation too.  And yes, I want folks to have to type the dreaded letters.

File: ci/cloudbuild/builds/shared.sh (hunk 3 above)
Comment (coryan) [reply ci/cloudbuild/builds/shared.sh]: Fixed.

File: ci/cloudbuild/builds/publish-docs.sh (hunk 4 above)
Comment (coryan) [reply ci/cloudbuild/builds/publish-docs.sh]: Fixed

File: ci/cloudbuild/builds/cmake-single-feature.sh (hunk 5 above)
Comment (coryan) [reply ci/cloudbuild/builds/cmake-single-feature.sh:55]: The loop below does too much.  It tries to verify `google_cloud_cpp_${feature}.pc` files have certain contents, that does not apply to `__ga_libraries__` and friends.

Comment (dbolduc) [reply ci/cloudbuild/builds/cmake-single-feature.sh:55]: D'oh. I somehow read this as looping over each library in the GA + experimental list. Not the strings "__ga_libraries__", "__experimental_libraries__". My bad.

File: cmake/GoogleCloudCppFeatures.cmake (hunk 2 above)
Comment (dbolduc) [reply cmake/GoogleCloudCppFeatures.cmake:148]: I see you opted not to document `__experimental_libraries__`. I guess that one is an easter egg.

**Review by dbolduc — APPROVED** [review]

**Pull request merged**